cluster-lifecycle/update-managedcluster-status-to-include-clusterinfo.md

```diff
@@ -0,0 +1,254 @@
+| ClusterID | the identifier of managed cluster | |
+| DistributionInfo | the information about distribution of managed cluster | **Not merged**. It's product specific. |
+| ConsoleURL | the url of console in managed cluster | ？ |
+| NodeList | a list of the status of nodes | **Not merged**. There is a scale issue for a cluster with a large number of nodes. |
+| LoggingEndpoint | the endpoint to connect to logging server of managed cluster | **Not merged**. It's product specific. |
+| LoggingPort | the port to connect to logging server of managed cluster | **Not merged**. It's product specific. |
+
+ClusterClaim is a CRD which comes from the upstream proposal for ClusterID. It’s a cluster-scoped resource created on managed clusters and provides a way to store cluster scoped information. Its name is a well known or custom name to identify the claim. The following shows the definition of ClusterClaim.
+
+```go
+type ClusterClaim struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the attributes of the ClusterClaim.
+    Spec ClusterClaimSpec `json:"spec,omitempty"`
+}
+
+type ClusterClaimSpec struct {
+    // Value is a claim-dependent string
+    Value string `json:"value,omitempty"`
+
+    // Signature may be used to validate a specific claim
+    // +optional
+    Signature string `json:"signature,omitempty"`
+}
+```
+
+An example claim for clusterID looks like,
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: ClusterClaim
+metadata:
+  name: id.k8s.io
+spec:
+  value: 95f91f25-d7a2-4fc3-9237-2ef633d8451c
+```
+
+ClusterClaims will be created on the managed cluster for the following fields of ClusterInfoStatus:
+
+| Field Name | Claim Name |
+| ---- | ---- |
+| KubeVendor | kubevendor.cluster.open-cluster-management.io |
+| CloudVendor | cloudvendor.cluster.open-cluster-management.io |
+| ClusterID | id.k8s.io |
+| ConsoleURL | consoleurl.cluster.open-cluster-management.io |
+
+And on hub side, add a list of ClusterClaims in ManagedClusterStatus.
+
+```go
+// ManagedClusterStatus represents the current status of joined managed cluster.
+type ManagedClusterStatus struct {
+    // Conditions contains the different condition statuses for this managed cluster.
+    Conditions []metav1.Condition `json:"conditions"`
+
+    // Capacity represents the total resource capacity from all nodeStatuses
+    // on the managed cluster.
+    Capacity ResourceList `json:"capacity,omitempty"`
+
+    // Allocatable represents the total allocatable resources on the managed cluster.
+    Allocatable ResourceList `json:"allocatable,omitempty"`
+
+    // Version represents the kubernetes version of the managed cluster.
+    Version ManagedClusterVersion `json:"version,omitempty"`
+
+    // ClusterClaims represents cluster information that a managed cluster claims
+    // +optional
+    ClusterClaims []ManagedClusterClaim `json:"clusterClaims,omitempty"`
```

Comment:
Oh, nice.  I think I like this.

Reply:
Do we want to qualify what would be in this list?

Reply:
In the current design, all ClusterClaims created on managed cluster will be synced into this slice. If necessary, it's possible to filter claims on managed cluster with a particular label or certain suffixes in name.

Reply:
maybe we could restrict that clusterclaims name with a suffix of `k8s.io` and `open-cluster-management.io` ?

Reply:
it's something to think about, i suspect that once this API is available you will have lot more use of it than you might expect. If there were a hundred claims, would we want to expose them all?

Reply:
How about to add a parameter `cluster-claims-max` for registration agent with default value `20`? The reserved claims (including id.k8s.io, kubeversion.cluster.open-cluster-management.io, platform.cluster.open-cluster-management.io and product.cluster.open-cluster-management.io) have high priority. Some of the customized claims might not be exposed once the total number of the ClusterClaims on a managed cluster exceeds the value of `cluster-claims-max`.